### PR TITLE
(build) Split project into gui and core

### PR DESCRIPTION
**Pull request by alediaferia**
This is the first step towards a more modular files structure.
The core target will contain only the core blockchain technology
in order to make it easier to link against when testing. The gui
project instead will be the whole Qt frontend.

As part of this PR I've also tried to fix a few tests as I went along. If you feel they should belong to a different PR I'll exclude them from this one. If this PR goes in I'll move on to enable the tests target too and re-enable testing for this project.

### CODE REVIEW

**Comment by ivangabriele** [comment]
Since there is some deep cleaning and many test to review, I think that it's not an issue to have "long" PRs as long as they only concern tests and refactoring.

**Comment by AlejandroVera** [comment]
Nice! Ii would be even better if you set Jenkins to run the tests @alediaferia !

File: src/core.pro [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+} else {
+    message(Building with UPNP support)
+    count(USE_UPNP, 0) {
+        USE_UPNP=1
+    }
+    DEFINES += USE_UPNP=$$USE_UPNP STATICLIB
+    INCLUDEPATH += $$MINIUPNPC_INCLUDE_PATH
+}
+
+# use: qmake "USE_IPV6=1" ( enabled by default; default)
```
Comment (greenbigfrog) [inline src/core.pro]: Extra space

File: src/core.pro [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+    QMAKE_CXXFLAGS *= -fstack-protector-all --param ssp-buffer-size=1
+    QMAKE_LFLAGS *= -fstack-protector-all --param ssp-buffer-size=1
+    # We need to exclude this for Windows cross compile with MinGW 4.2.x, as it will result in a non-working executable!
+    # This can be enabled for Windows, when we switch to MinGW >= 4.4.x.
+}
+# for extra security on Windows: enable ASLR and DEP via GCC linker flags
+win32:QMAKE_LFLAGS *= -Wl,--large-address-aware -static
+win32:QMAKE_LFLAGS += -static-libgcc -static-libstdc++
+
+# use: qmake "USE_UPNP=1" ( enabled by default; default)
```
Comment (greenbigfrog) [inline src/core.pro]: Extra space

File: src/gui.pro [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+}
+
+# use: qmake "USE_DBUS=1"
+contains(USE_DBUS, 1) {
+    message(Building with DBUS (Freedesktop notifications) support)
+    DEFINES += USE_DBUS
+    QT += dbus
+}
+
+# use: qmake "USE_IPV6=1" ( enabled by default; default)
```
Comment (greenbigfrog) [inline src/gui.pro]: Extra space

File: src/leveldb.pri [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+!win32 {
+    # we use QMAKE_CXXFLAGS_RELEASE even without RELEASE=1 because we use RELEASE to indicate linking preferences not -O preferences
+    genleveldb.commands = cd $$PWD/leveldb && CC=$$QMAKE_CC CXX=$$QMAKE_CXX $(MAKE) OPT=\"$$QMAKE_CXXFLAGS $$QMAKE_CXXFLAGS_RELEASE\" libleveldb.a libmemenv.a
+} else {
+    # make an educated guess about what the ranlib command is called
+    isEmpty(QMAKE_RANLIB) {
+        QMAKE_RANLIB = $$replace(QMAKE_STRIP, strip, ranlib)
+    }
+    LIBS += -lshlwapi
+    #genleveldb.commands = cd $$PWD/leveldb && CC=$$QMAKE_CC CXX=$$QMAKE_CXX TARGET_OS=OS_WINDOWS_CROSSCOMPILE $(MAKE) OPT=\"$$QMAKE_CXXFLAGS $$QMAKE_CXXFLAGS_RELEASE\" libleveldb.a libmemenv.a && $$QMAKE_RANLIB $$PWD/leveldb/libleveldb.a && $$QMAKE_RANLIB $$PWD/leveldb/libmemenv.a
```
Comment (greenbigfrog) [inline src/leveldb.pri]: Missing space. Also, does this have to stay here?

File: src/leveldb.pri [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+SOURCES += $$PWD/txdb-leveldb.cpp \
+    $$PWD/bloom.cpp \
+    $$PWD/hash.cpp \
+    $$PWD/blake.c \
+    $$PWD/groestl.c \
+    $$PWD/jh.c \
+    $$PWD/keccak.c \
+    $$PWD/skein.c
+!win32 {
+    # we use QMAKE_CXXFLAGS_RELEASE even without RELEASE=1 because we use RELEASE to indicate linking preferences not -O preferences
```
Comment (greenbigfrog) [inline src/leveldb.pri]: Capitalize `We`

File: src/leveldb.pri [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    genleveldb.commands = cd $$PWD/leveldb && CC=$$QMAKE_CC CXX=$$QMAKE_CXX $(MAKE) OPT=\"$$QMAKE_CXXFLAGS $$QMAKE_CXXFLAGS_RELEASE\" libleveldb.a libmemenv.a
+} else {
+    # make an educated guess about what the ranlib command is called
```
Comment (greenbigfrog) [inline src/leveldb.pri]: Capitalize `Make`

File: src/test/wallet_tests.cpp [hunk 7]
```diff
@@ -1,5 +1,7 @@
 #include <boost/test/unit_test.hpp>
 
+#include <inttypes.h>
+
```
Comment (greenbigfrog) [inline src/test/wallet_tests.cpp]: Why the extra newline here?

**Comment by Hyjaz** [comment]
@alediaferia What are we waiting to merge this pull request?

**Comment by alediaferia** [comment]
@AlejandroVera will do as soon as I manage to create a makefile for them too. Should be easier after this change.

**Comment by alediaferia** [comment]
@Hyjaz nothing really. I think it's had enough visibility. Will merge .

**Comment by greenbigfrog** [comment]
Why did this get merged without applying the corrections?

**Comment by alediaferia** [comment]
Hi @greenbigfrog ,
if you're talking about the stylistic corrections I think that we haven't agreed on a coding style anywhere and we should put some effort into it. For the time being, I've just moved already existing code around to split the build. I appreciate your comments but those style issues are allover the place in the whole codebase and there's not much consistency.
We should first decide on which coding style to follow. Do you have any recommendation on how to approach this problem? We could pick a sensible coding style and document it clearly in the project that from that point on we will follow that and include automated checks in our PR hooks.
Let me know.

**Comment by greenbigfrog** [comment]
Of course documenting the coding style somewhere is the best way to go. From my POV though, you have to start somewhere when changing/improving the coding style, so you might as well have edited them right away 😄

**Pull request merged**